ipa-core/src/hpke/registry.rs

```diff
@@ -58,10 +58,28 @@ impl Deref for PublicKeyOnly {
     }
 }
 
+// This newtype is necessary because IpaPrivateKey is an associated type from another crate (hpke).
+// The coherence rules prohibit us from implementing `PrivateKeyRegistry` both for our concrete type
+// `KeyPair` and for `IpaPrivateKey`, because the impls would overlap if hpke chose to define
+// `IpaPrivateKey` to be the same as `KeyPair`.
+pub struct PrivateKeyOnly(pub IpaPrivateKey);
+
+impl Deref for PrivateKeyOnly {
+    type Target = IpaPrivateKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
 pub trait PublicKeyRegistry {
     fn public_key(&self, key_id: KeyIdentifier) -> Option<&IpaPublicKey>;
 }
 
+pub trait PrivateKeyRegistry: Send + Sync + 'static {
```

Comment:
Now that you're using generics, some of these bounds may no longer be required.

Reply:
good catch. I removed the function bounds elsewhere, so this should be able to go.

Reply:
Ok, nevermind. Looks like they are still all required:
Sync and Send:
```
error[E0277]: `R` cannot be sent between threads safely
   --> ipa-core/src/query/executor.rs:107:17
    |
107 | /                 Box::pin(
108 | |                     OprfIpaQuery::<BA16, R>::new(ipa_config, key_registry)
109 | |                         .execute(ctx, config.size, input)
110 | |                         .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
111 | |                 )
    | |_________________^ `R` cannot be sent between threads safely
    |
    = note: required for `Arc<R>` to implement `std::marker::Send`
note: required because it appears within the type `OprfIpaQuery<'_, boolean_array_16::BA16, R>`
   --> ipa-core/src/query/runner/oprf_ipa.rs:32:12
    |
32  | pub struct OprfIpaQuery<'a, HV, R: PrivateKeyRegistry> {
    |            ^^^^^^^^^^^^
note: required because it's used within this `async` fn body
   --> ipa-core/src/query/runner/oprf_ipa.rs:57:5
    |
57  |     #[tracing::instrument("oprf_ipa_query", skip_all, fields(sz=%query_size))]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required because it appears within the type `Map<impl Future<Output = Result<Vec<AdditiveShare<BA16>>, Error>>, {closure@executor.rs:110:31}>`
   --> /Users/eriktaubeneck/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/map.rs:15:14
    |
15  |     pub enum Map<Fut, F> {
    |              ^^^
note: required because it appears within the type `Map<impl Future<Output = Result<Vec<AdditiveShare<BA16>>, Error>>, {closure@executor.rs:110:31}>`
   --> /Users/eriktaubeneck/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/mod.rs:50:5
    |
50  |     Map<Fut, F>(
    |     ^^^
note: required because it appears within the type `Flatten<Map<impl Future<Output = Result<Vec<AdditiveShare<BA16>>, Error>>, {closure@executor.rs:110:31}>, Ready<Result<Box<...>, ...>>>`
   --> /Users/eriktaubeneck/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/flatten.rs:13:14
    |
13  |     pub enum Flatten<Fut1, Fut2> {
    |              ^^^^^^^
note: required because it appears within the type `Then<impl Future<Output = Result<Vec<AdditiveShare<BA16>>, Error>>, Ready<Result<Box<dyn Result>, Error>>, {closure@executor.rs:110:31}>`
   --> /Users/eriktaubeneck/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/mod.rs:71:5
    |
71  |     Then<Fut1, Fut2, F>(
    |     ^^^^
    = note: required for the cast from `Pin<Box<Then<impl Future<Output = Result<Vec<AdditiveShare<BA16>>, Error>>, Ready<Result<Box<dyn Result>, Error>>, ...>>>` to `Pin<Box<dyn futures::Future<Output = std::result::Result<Box<(dyn query::executor::Result + 'static)>, error::Error>> + std::marker::Send>>`
help: consider further restricting this bound
    |
63  | pub fn execute<R: PrivateKeyRegistry + std::marker::Send>(
help: consider further restricting this bound
    |
63  | pub fn execute<R: PrivateKeyRegistry + std::marker::Sync>(
```

`static:
```
error[E0310]: the parameter type `R` may not live long enough
   --> ipa-core/src/query/executor.rs:107:17
    |
107 | /                 Box::pin(
108 | |                     OprfIpaQuery::<BA16, R>::new(ipa_config, key_registry)
109 | |                         .execute(ctx, config.size, input)
110 | |                         .then(|res| ready(res.map(|out| Box::new(out) as Box<dyn Result>))),
111 | |                 )
    | |                 ^
    | |                 |
    | |_________________the parameter type `R` must be valid for the static lifetime...
    |                   ...so that the type `R` will meet its required lifetime bounds
    |
help: consider adding an explicit lifetime bound
    |
63  | pub fn execute<R: PrivateKeyRegistry + 'static>(
```